packages/flutter_tools/lib/src/features.dart

```diff
@@ -0,0 +1,245 @@
+// found in the LICENSE file.
+
+import 'package:meta/meta.dart';
+
+import 'base/config.dart';
+import 'base/context.dart';
+import 'base/platform.dart';
+import 'version.dart';
+
+/// The current [FeatureFlags] implementation.
+///
+/// If not injected, a default implementation is provided.
+FeatureFlags get featureFlags => context.get<FeatureFlags>() ?? const FeatureFlags();
+
+/// The interface used to determine if a particular [Feature] is enabled.
+///
+/// The rest of the tools code should use this class instead of looking up
+/// features directly. To faciliate rolls to google3 and other clients, all
+/// flags should be provided with a default implementation here. Clients that
+/// use this class should extent instead of implement, so that new flags are
+/// picked up automatically.
+class FeatureFlags {
+  const FeatureFlags();
+
+  /// Whether flutter web is enabled.
+  ///
+  /// Defaults to `false`.
+  bool get isWebEnabled => false;
+
+  /// Whether flutter desktop for macOS is enabled.
+  ///
+  /// Defaults to `false`.
+  bool get isMacOSEnabled => false;
+
+  /// Whether flutter desktop for Windows is enabled.
+  ///
+  /// Defaults to `false`.
+  bool get isWindowsEnabled => false;
+
+  /// Whether flutter desktop for linux is enabled.
+  ///
+  /// Defaults to `false`.
+  bool get isLinuxEnabled => false;
+}
+
+/// All current Flutter feature flags.
+const List<Feature> allFeatures = <Feature>[
+  flutterWebFeature,
+  flutterLinuxDesktopFeature,
+  flutterMacOSDesktopFeature,
+  flutterWindowsDesktopFeature,
+];
+
+/// The [Feature] for flutter web.
+const Feature flutterWebFeature = Feature(
+  name: 'Flutter Web',
+  configSetting: 'enable-web',
+  environmentOverride: 'FLUTTER_WEB',
+  master: FeatureSetting(
+    available: true,
+    enabledByDefault: false,
+  ),
+  dev: FeatureSetting(
+    available: true,
+    enabledByDefault: false,
+  ),
+);
+
+/// The [Feature] for macOS desktop.
+const Feature flutterMacOSDesktopFeature = Feature(
+  name: 'Flutter Desktop for macOS',
+  configSetting: 'enable-macos-desktop',
+  environmentOverride: 'ENABLE_FLUTTER_DESKTOP',
+  master: FeatureSetting(
+    available: true,
+    enabledByDefault: false,
+  ),
+);
+
+/// The [Feature] for Linux desktop.
+const Feature flutterLinuxDesktopFeature = Feature(
+  name: 'Flutter Desktop for Linux',
+  configSetting: 'enable-linux-desktop',
+  environmentOverride: 'ENABLE_FLUTTER_DESKTOP',
+  master: FeatureSetting(
+    available: true,
+    enabledByDefault: false,
+  ),
+);
+
+/// The [Feature] for Windows desktop.
+const Feature flutterWindowsDesktopFeature = Feature(
+  name: 'Flutter Desktop for Windows',
+  configSetting: 'enable-windows-desktop',
+  environmentOverride: 'ENABLE_FLUTTER_DESKTOP',
+  master: FeatureSetting(
+    available: true,
+    enabledByDefault: false,
+  ),
+);
+
+/// A [FeatureFlags] that looks up values based on [Feature] definitions.
+class FlutterFeatureFlags implements FeatureFlags {
```

Comment:
This class name doesn't add anything to FeatureFlags :)

Perhaps ConfigBasedFeatureFlags?

Reply:
Good point, removed for now. google3 should be unaffected because we do not use the feature flags yet.